### ISSUE TITLE
Is there any function like torch.autograd.grad() in oneflow?

### ISSUE THREAD

**Issue by 15168387900** [Oneflow-Inc/oneflow#3801]
Is there a function like torch.autograd.grad() in oneflow?

**Comment by MARD1NO** [comment]
Thanks for your feedback. Now the `autograd` in OneFlow still in developing.

**Comment by 15168387900** [comment]
> Thanks for your feedback. Now the `autograd` in OneFlow still in developing.
> 
> However, if you want to get the grad of variable, it's recommend you to use `flow.watch_diff`
> 
> you can refer to [watch_diff](https://docs.oneflow.org/extended_topics/watch_watch_diff.html#watch_diff)

Thank you for replying.
 I am not very clear about the usage of watch_diff. 
Can you give me a detailed introduction?
I need to get the gradient of a Blob.
I want to add the gradient to the loss in my model.

**Comment by doombeaker** [comment]
I'm sorry what we seldom meet the situation that we need to "get the gradient and add it to the loss after processing".
So it may be very tricky(if not impossible), for example, use two-stage job function, to do that.

Can you describe more details about your work?(For example, attach the code implemented by other framework).

Thank you for your attention and look forward to your reply.

**Comment by 15168387900** [comment]
> I'm sorry what we seldom meet the situation that we need to "get the gradient and add it to the loss after processing".
> So it may be very tricky(if not impossible), for example, use two-stage job function, to do that.
> 
> Can you describe more details about your work?(For example, attach the code implemented by other framework).
> 
> Thank you for your attention and look forward to your reply.

the code implemented by other framework just like this:

import torch

gradients = torch.autograd.grad(outputs=d_hat.sum(), inputs=hat, retain_graph=True, create_graph=True, only_inputs=True)[0]
gradients = gradients.view(gradients.size(0), -1)
gradient_norm = gradients.norm(2, dim=1)
gradient_penalty = 10 * gradient_norm.sub(1).pow(2).mean()
loss_d += gradient_penalty

**Comment by Ldpe2G** [comment]
@15168387900  oneflow doesn't provide api like `torch.autograd.grad()` yet. 
Maybe you can try the approach described below :
[生成式对抗网络GAN有哪些最新的发展，可以实际应用到哪些场景中？](https://www.zhihu.com/question/52602529/answer/158727900)

To use difference instread of differential to approximate the gradient penalty.
